0.4.7 - Updated repository to use 'uv' + resolved dependency errors.

NeilGraham opened the pull request:
## **User description**
## Changes made

- Update repository to support [uv](https://github.com/astral-sh/uv):
  - Replaced **setup.py** with **pyproject.toml** definitions.
- Resolved **pyoxigraph** dependency errors:
  - Using `pyoxigraph.RdfFormat.*` objects instead of using mime type strings when parsing and serializing RDF.
- Resolved **rdflib** dependency errors:
  - Now using `rdflib.Dataset` for all references to now deprecated `rdflib.ConjunctiveGraph`.
- Incremented PyPi version from 0.4.6 to 0.4.7

## Testing the Repository

After setting up [uv](https://github.com/astral-sh/uv), you can run the following:

```bash
cd rdfhash
uv sync
uv run pytest
```

@sa-bpelakh, @Jamie-SA & @dhurlburtusa: 
If either of you have some time, could you help review these changes?
Let me know if you have any repository related questions.


___

## **CodeAnt-AI Description**
Support standard RDF media types, multiple graph backends, and modern packaging

### What Changed
- CLI now accepts common RDF format names and file extensions and validates them, producing clear errors when an unsupported format or accept value is provided.
- Graph handling: media types and file extension detection use the system mimetypes table; rdflib now uses Dataset-based graphs and pyoxigraph I/O uses proper RdfFormat objects so parsing and serialization work with current library versions and triple-vs-quad formats.
- Packaging and developer workflows updated: removed setup.py in favor of pyproject.toml with hatchling, bumped package version, and README adds pip and UV usage instructions for installing and testing the project.

### Impact
`✅ Clearer RDF format errors in CLI`
`✅ Fewer parsing/serialization failures with modern RDF libraries`
`✅ Easier developer install and test with pip or UV`
<details>
<summary><strong>💡 Usage Guide</strong></summary>

### Checking Your Pull Request
Every time you make a pull request, our system automatically looks through it. We check for security issues, mistakes in how you're setting up your infrastructure, and common code problems. We do this to make sure your changes are solid and won't cause any trouble later.

### Talking to CodeAnt AI
Got a question or need a hand with something in your pull request? You can easily get in touch with CodeAnt AI right here. Just type the following in a comment on your pull request, and replace "Your question here" with whatever you want to ask:
<pre>
<code>@codeant-ai ask: Your question here</code>
</pre>
This lets you have a chat with CodeAnt AI about your pull request, making it easier to understand and improve your code.

#### Example
<pre>
<code>@codeant-ai ask: Can you suggest a safer alternative to storing this secret?</code>
</pre>

### Preserve Org Learnings with CodeAnt
You can record team preferences so CodeAnt AI applies them in future reviews. Reply directly to the specific CodeAnt AI suggestion (in the same thread) and replace "Your feedback here" with your input:
<pre>
<code>@codeant-ai: Your feedback here</code>
</pre>
This helps CodeAnt AI learn and adapt to your team's coding style and standards.

#### Example
<pre>
<code>@codeant-ai: Do not flag unused imports.</code>
</pre>

### Retrigger review
Ask CodeAnt AI to review the PR again, by typing:
<pre>
<code>@codeant-ai: review</code>
</pre>

### Check Your Repository Health
To analyze the health of your code repository, visit our dashboard at [https://app.codeant.ai](https://app.codeant.ai). This tool helps you identify potential issues and areas for improvement in your codebase, ensuring your repository maintains high standards of code health.

</details>

NeilGraham committed:
Resolved 'pyoxigraph' dependency errors.
Update repository to use 'uv'.
Replaced setup.py with 'pyproject.toml' definitions.

dhurlburtusa commented on rdfhash/utils/__init__.py:
```diff
@@ -2,6 +2,37 @@
 
 from .hash import hash_types
 
+triple_mime_types = {
```
MIME type is an outdated term. I'd use media type. The `pyoxigraph` library is even using media type as indicated by the `from_media_type` function.

NeilGraham replied:
Renamed all references to 'MIME type' to now reference 'media type'. 
It is funny how Python still has `mimetypes` as a standard library though!

dhurlburtusa commented on rdfhash/utils/__init__.py:
```diff
@@ -2,6 +2,37 @@
+}
+
+
+def get_rdf_mime_type(format: str):
```
Are you expecting the first argument to be a file extension, a media type, or either? I'm not sure what is expected based on the parameter name of `format`.

sa-bpelakh replied:
Yea, that variable should be named something like `file_ext`

NeilGraham replied:
This primarily takes a --format argument with an alias value and returns it's media type (e.g. `turtle` -> `text/turtle`). 

I renamed it to `def get_rdf_media_type(format:str)`

sa-bpelakh commented on pyproject.toml:
```diff
@@ -1,6 +1,45 @@
+]
+
+[project.urls]
+Homepage = "https://github.com/NeilGraham/rdfhash"
```
Should these not point to this copy of the repo, as opposed to your private copy?

NeilGraham replied:
Updated all references to `https://github.com/NeilGraham/rdfhash` to now point to `https://github.com/semanticarts/rdfhash`

sa-bpelakh commented on rdfhash/utils/__init__.py:
```diff
@@ -2,6 +2,37 @@
+    "nq": "application/n-quads",
+    "rdf": "application/rdf+xml",
+}
+rdf_mime_types = {
```
This is just `rdf_ext_to_mime_types.values()`, there is no need to maintain a separate list.

NeilGraham replied:
Removed all redundancies regarding media types, now using `mimetypes` library to define media type to file extension pairings and just maintaining the following objects in `rdfhash/utils/__init__.py` for storing supported rdf media types and alias mappings:

```python
import mimetypes

# ...

mimetypes.add_type("text/turtle", ".ttl")
mimetypes.add_type("application/n-triples", ".nt")
mimetypes.add_type("application/trig", ".trig")
mimetypes.add_type("application/n-quads", ".nq")
mimetypes.add_type("application/rdf+xml", ".rdf")
mimetypes.add_type("text/n3", ".n3")

#: Supported RDF media types.
rdf_media_types = {
    "text/turtle",
    "application/n-triples",
    "application/trig",
    "application/n-quads",
    "application/rdf+xml",
    "text/n3",
}

#: Triple-based RDF media types.
triple_media_types = {
    "text/turtle",
    "application/n-triples",
}

#: RDF file type names to RDF media types.
rdf_alias_to_media_type = {
    "nquads": "application/n-quads",
    "ntriples": "application/n-triples",
    "turtle": "text/turtle",
    "xml": "application/rdf+xml",
    # Associate file extensions (without the leading dot) with RDF media types.
    **{mimetypes.guess_extension(mt)[1:]: mt for mt in rdf_media_types},
}
```

sa-bpelakh commented on rdfhash/utils/__init__.py:
```diff
@@ -2,6 +2,37 @@
+            "Invalid RDF format: {format}. Must be one of {rdf_mime_types}.".format(
+                format=format, rdf_mime_types=rdf_mime_types
+            )
+        )
```
This doesn't appear to return `mime_type` when successful. Code should be something like
```python
     try:
         return rdf_ext_to_mime_types[format.lower()]
     except KeyError:
        raise ValueError(
            f"Unrecognized file extension: {format}. Must be one of {rdf_mime_types}."
        )
```

NeilGraham replied:
Now returning media type:

```python
def get_rdf_media_type(format: str):
    """Convert a simplified RDF format string or media type to a supported RDF media type.

    Args:
        format (str): Simplified string or media type.

    Raises:
        ValueError: If 'format' is not related to a supported RDF media type.
    """
    format = format.lower()
    media_type = rdf_alias_to_media_type.get(format, format)
    if media_type not in rdf_media_types:
        raise ValueError(
            f"Invalid RDF format: {format}. Must be one of {rdf_media_types}."
        )
    return media_type
```

sa-bpelakh commented on rdfhash/utils/graph.py:
```diff
@@ -287,6 +292,22 @@ class OxiGraph(__Graph__):
+            "application/ld+json": pyoxigraph.RdfFormat.JSON_LD,
+        }
+        return format_map.get(
+            format_str, pyoxigraph.RdfFormat.from_media_type(format_str)
```
If `pyoxigraph` already provides the `from_media_type` method, are you overriding some of the choices?

NeilGraham replied:
Removed redundant mapping and just relying on `pyoxigraph.RdfFormat.from_media_type(media_type)`:

```python
    @staticmethod
    def _get_rdf_media_type(format_str):
        """Convert media type string to pyoxigraph.RdfFormat object"""
        return pyoxigraph.RdfFormat.from_media_type(get_rdf_media_type(format_str))
```

sa-bpelakh commented on uv.lock:
```diff
@@ -0,0 +1,630 @@
+version = 1
```
Should this file be under CM? Isn't it auto-generated from `pyproject.toml` by `uv`?

NeilGraham replied:
It is auto-generated by uv, but it is still recommended to commit this file to version control: 
https://docs.astral.sh/uv/guides/projects/#uvlock

> [uv.lock] should be checked into version control, allowing for consistent and reproducible installations across machines.

sa-bpelakh reviewed (commented):
Could use a little cleanup

NeilGraham committed:
Updated repository with respect to reviews.
Using 'media type' in place of 'mime type' across repo.
Updated 'rdfhash/utils/__init__.py' to use 'mimetypes' library for extending media type to extension matching using 'mimetypes.guess_type(file)'.
Simplified CLI arguments --format and --accept to make use of 'get_rdf_media_type' for resolving aliases to media types.
Updated pyproject.toml to define developer dependencies under 'dependency-groups.dev'.
Added comments throughout project.

NeilGraham commented on rdfhash/utils/graph.py:
```diff
@@ -297,22 +273,28 @@ def __contains__(self, item):
 
     def _parse_file(self, path, format):
-        self.graph.load(path, format)
+        self.graph.load(path=path, format=self._get_rdf_media_type(format))
```
First positional argument on `pyoxigraph.Store.load` is `input` expecting file content, not `path` expecting a file path. 👎

NeilGraham committed:
Resolved suggestions made by CodeAnt AI

NeilGraham commented on rdfhash/cli.py:
```diff
@@ -109,27 +106,22 @@ def cli(args_list=None):
     parser = get_parser()
     args = parser.parse_args(["--help"] if len(args_list) == 0 else sys.argv[1:])
```
Resolved this issue by passing args_list to `parser.parse_args`
